File: leetcode/Python/different-ways-to-add-parentheses.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        dp, nums, ops = [[[] for j in range(10)] for i in range(10)], [] , []
        def _parsing():
            s = []
            for ch in expression:
                if ch>='0' and ch<='9':
                    s.append(ch)
                else:
                    nums.append(int(''.join(s)))
                    ops.append(ch)
                    s.clear()
            nums.append(int(''.join(s)))
        def _subCalculation(l, r):
            t=0
            if l==r:
                dp[l][r].append(nums[l])
            for i in range(l,r):
                for a in dp[l][i]:
                    for b in dp[i+1][r]:
                        if ops[i]=='+':
                            t=a+b
                        elif ops[i]=='-':
                            t=a-b
                        else:
                            t=a*b
                        dp[l][r].append(t)
        _parsing()
        for i in range(len(nums)):
            l, r = 0, i
            while r<len(nums):
                _subCalculation(l,r)
                l, r = l+1, r+1
        return dp[0][len(nums)-1]

File: leetcode/C++/different-ways-to-add-parentheses.cpp
class Solution {
public:
    vector<int> nums, op;
    vector<int> dp[10][10];
    void parsing(string &expression){
        string s;
        for(char &ch : expression){
            if(ch>='0' && ch<='9') s.push_back(ch);
            else {
                nums.push_back(stoi(s)); s="";
                if(ch=='+') op.push_back(0);
                else if(ch=='-') op.push_back(1);
                else op.push_back(2);
            }
        }
        nums.push_back(stoi(s));
    }
    void subOperation(int &l, int &r){
        if(l==r) dp[l][r].push_back(nums[l]);
        for(int i=l,t=0; i<r; ++i){
            for(int &a : dp[l][i]) {
                for(int &b : dp[i+1][r]){
                    if(op[i]==0) t=a+b;
                    else if(op[i]==1) t=a-b;
                    else t=a*b;
                    dp[l][r].push_back(t);
                }
            }
        }
    }
    vector<int> diffWaysToCompute(string expression) {
        parsing(expression);
        for(int i=0; i<nums.size(); ++i)
            for(int l=0,r=i; r<nums.size(); ++l,++r)
                subOperation(l,r);
        return dp[0][nums.size()-1];
    }
};